File: Java/src/main/java/Solution401_500/Solution463.java
package Solution401_500;

/**
 * @author windows
 */
public class Solution463 {
    public int islandPerimeter(int[][] grid) {
        int sum = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == 1){
                    sum += 4;
                    if (i - 1 >= 0 && grid[i - 1][j] == 1){
                        sum--;
                    }
                    if (i + 1 < grid.length && grid[i + 1][j] == 1){
                        sum--;
                    }
                    if (j - 1 >= 0 && grid[i][j - 1] == 1){
                        sum--;
                    }
                    if (j + 1 < grid[i].length && grid[i][j + 1] == 1){
                        sum--;
                    }
                }
            }
        }
        return sum;
    }


    public static void main(String[] args) {
        Solution463 solution463 = new Solution463();
        System.out.println(solution463.islandPerimeter(new int[][]{{0, 1, 0, 0}, {1, 1, 1, 0}, {0, 1, 0, 0}, {1, 1, 0, 0}}));
    }
}

File: C++/src/Solution401_500/Solution463.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution463.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;

int Solution463::islandPerimeter(vector<vector<int>> &grid) {
    int res = 0;
    int m = grid.size();
    int n = grid[0].size();
    constexpr static int dx[4] = {0 , 1, 0, - 1};
    constexpr static int dy[4] = {1 , 0, -1, 0};
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (grid[i][j]){
                for (int k = 0; k < 4; ++k) {
                    int x = i + dx[k];
                    int y = j + dy[k];
                    if (x < 0 || y < 0 || x >= m || y >= n || !grid[x][y]){
                        res++;
                    }
                }
            }
        }
    }
    return res;
}
